presto-thrift/src/main/java/com/facebook/presto/thrift/ThriftMetadata.java

```diff
@@ -0,0 +1,261 @@
+
+public class ThriftMetadata
+        implements ConnectorMetadata
+{
+    private static final Duration EXPIRE_AFTER_WRITE = new Duration(10, MINUTES);
+    private static final Duration REFRESH_AFTER_WRITE = new Duration(2, MINUTES);
+
+    private final PrestoClientProvider clientProvider;
+    private final ThriftClientSessionProperties clientSessionProperties;
+    private final LoadingCache<SchemaTableName, Optional<ConnectorTableMetadata>> tableCache;
+
+    @Inject
+    public ThriftMetadata(
+            PrestoClientProvider clientProvider,
+            TypeManager typeManager,
+            ThriftClientSessionProperties clientSessionProperties,
+            @ForMetadataRefresh Executor metadataRefreshExecutor)
+    {
+        this.clientProvider = requireNonNull(clientProvider, "clientProvider is null");
+        requireNonNull(typeManager, "typeManager is null");
+        this.clientSessionProperties = requireNonNull(clientSessionProperties, "clientSessionProperties is null");
+        this.tableCache = CacheBuilder.newBuilder()
+                .expireAfterWrite(EXPIRE_AFTER_WRITE.toMillis(), MILLISECONDS)
+                .refreshAfterWrite(REFRESH_AFTER_WRITE.toMillis(), MILLISECONDS)
+                .build(asyncReloading(new CacheLoader<SchemaTableName, Optional<ConnectorTableMetadata>>()
+                {
+                    @Override
+                    public Optional<ConnectorTableMetadata> load(SchemaTableName schemaTableName)
+                            throws Exception
+                    {
+                        requireNonNull(schemaTableName, "schemaTableName is null");
+                        try (ThriftPrestoClient client = clientProvider.connectToAnyHost()) {
+                            ThriftNullableTableMetadata thriftTableMetadata = client.getTableMetadata(fromSchemaTableName(schemaTableName));
+                            if (thriftTableMetadata.getThriftTableMetadata() == null) {
+                                return Optional.empty();
+                            }
+                            else {
+                                return Optional.of(toConnectorTableMetadata(thriftTableMetadata.getThriftTableMetadata(), typeManager));
+                            }
+                        }
+                    }
+                }, metadataRefreshExecutor));
+    }
+
+    @Override
+    public List<String> listSchemaNames(ConnectorSession session)
+    {
+        try (ThriftPrestoClient client = clientProvider.connectToAnyHost()) {
+            return client.listSchemaNames();
+        }
+    }
+
+    @Override
+    public ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
+    {
+        Optional<ConnectorTableMetadata> tableMetadata = tableCache.getUnchecked(tableName);
+        if (!tableMetadata.isPresent()) {
+            return null;
+        }
+        else {
+            SchemaTableName actualTableName = tableMetadata.get().getTable();
```

Comment:
Because it's not the one that was given as a method parameter. It should be the same though. I added a `checkState` check when loading the cache.